server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -246,10 +258,132 @@ protected void reroute(ShardId shardId, String reason) {
             this.client = client;
         }
 
+        // node grouped shard requests
+        private final Map<DiscoveryNode, Map<ShardId, ShardRequestInfo>> queuedRequests = new ConcurrentHashMap<>();
+        // shard to nodes requests
+        private final Map<ShardId, Map<String, ActionListener<TransportNodesListGatewayStartedShards.NodesGatewayStartedShards>>>
+            queuedListeners = new ConcurrentHashMap<>();
+
+        Lister<NodesGatewayStartedShards, NodeGatewayStartedShards> lister =
+            new Lister<>() {
+                @Override
+                public void list(ShardId shardId, String customDataPath, DiscoveryNode[] nodes,
+                                 ActionListener<NodesGatewayStartedShards> listener) {
+                    // group shards by node
+                    for (DiscoveryNode node : nodes) {
+                        Map<ShardId, ShardRequestInfo> nodeLevelRequest = queuedRequests.computeIfAbsent(node, n -> new HashMap<>());
+                        nodeLevelRequest.put(shardId, new ShardRequestInfo(shardId, customDataPath, listener));
+                    }
+                }
+
+                @Override
+                public void flush() {
+                    Map<ShardId, String> shards = new HashMap<>();
+                    for (DiscoveryNode node : queuedRequests.keySet()) {
+                        Map<ShardId, ShardRequestInfo> requestMap = queuedRequests.get(node);
+                        for (ShardId shardId : requestMap.keySet()) {
+                            ShardRequestInfo shardRequest = requestMap.get(shardId);
+                            shards.put(shardRequest.shardId(), shardRequest.getCustomDataPath());
+                            if (node.getVersion().before(Version.V_7_16_0)) {
```

Comment:
The version in `master` is now 8.1.0; it's unlikely we'll backport this to an earlier version.